substrate/frame/staking-async/src/migrations/v18.rs

```diff
@@ -0,0 +1,408 @@
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Staking Async Pallet migration from v17 to v18.
+
+use crate::{
+	migrations::PALLET_MIGRATIONS_ID, pallet::pallet::ElectableStashes, weights::WeightInfo,
+	BalanceOf, Config, Ledger, Pallet, StakingLedger, UnlockChunk,
+};
+use alloc::{collections::BTreeMap, vec::Vec};
+use codec::{Decode, Encode};
+use core::fmt::Debug;
+use frame_support::{
+	migrations::{MigrationId, SteppedMigration, SteppedMigrationError},
+	pallet_prelude::*,
+	weights::WeightMeter,
+};
+
+pub(crate) mod v17 {
+	use crate::{BalanceOf, Config, Pallet};
+	use codec::{HasCompact, MaxEncodedLen};
+	use frame_support::{pallet_prelude::*, storage_alias};
+	use sp_staking::EraIndex;
+
+	#[derive(
+		PartialEq, Eq, Clone, Encode, Decode, DecodeWithMemTracking, Debug, TypeInfo, MaxEncodedLen,
+	)]
+	pub struct UnlockChunk<Balance: HasCompact + MaxEncodedLen> {
+		/// Amount of funds to be unlocked.
+		#[codec(compact)]
+		pub(crate) value: Balance,
+		/// Era number at which point it'll be unlocked.
+		#[codec(compact)]
+		pub(crate) era: EraIndex,
+	}
+
+	#[derive(
+		PartialEqNoBound,
+		EqNoBound,
+		CloneNoBound,
+		Encode,
+		Decode,
+		DebugNoBound,
+		TypeInfo,
+		MaxEncodedLen,
+	)]
+	#[scale_info(skip_type_params(T))]
+	pub struct StakingLedger<T: Config> {
+		pub stash: T::AccountId,
+		#[codec(compact)]
+		pub total: BalanceOf<T>,
+		#[codec(compact)]
+		pub active: BalanceOf<T>,
+		pub unlocking: BoundedVec<UnlockChunk<BalanceOf<T>>, T::MaxUnlockingChunks>,
+		#[codec(skip)]
+		pub(crate) controller: Option<T::AccountId>,
+	}
+
+	#[storage_alias]
+	pub type Ledger<T: Config> = StorageMap<
+		Pallet<T>,
+		Blake2_128Concat,
+		<T as frame_system::Config>::AccountId,
+		StakingLedger<T>,
+	>;
+
+	#[storage_alias]
+	pub type ElectableStashes<T: Config> = StorageValue<
+		Pallet<T>,
+		BoundedBTreeSet<<T as frame_system::Config>::AccountId, <T as Config>::MaxValidatorSet>,
+		ValueQuery,
+	>;
+}
+
+/// Operations to be performed during this migration.
+#[derive(
+	PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, scale_info::TypeInfo, MaxEncodedLen,
+)]
+pub enum MigrationSteps<T: Config> {
+	/// Migrate staking ledger storage. The cursor indicates the last processed account.
+	/// If None, start from the beginning.
+	MigrateStakingLedger { cursor: Option<T::AccountId> },
+	/// Migrate electable stashes storage.
+	MigrateElectableStashes,
+	/// Changes the storage version to 18.
+	ChangeStorageVersion,
+	/// No more operations to be performed.
+	Noop,
+}
+
+pub struct LazyMigrationV17ToV18<T: Config>(PhantomData<T>);
```

Comment:
Looks solid - thanks. Do you have an estimate of how many blocks this migration will take on Polkadot AH, based on current ledger size?